design/src/rfcs/rfc0039_forward_compatible_errors.md

```diff
@@ -0,0 +1,131 @@
+    Unhandled(Unhandled),
+}
+```
+
+- **Modeled Error**: An error with an named variant, e.g. `NoSuchUpload` above
+- **Unmodeled Error**: Any other error, e.g. if the server returned `ValidationException` for the above operation.
+- **Error code**: All errors across all protocols provide a `code`, a unique method to identify an error across the service closure.
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+In the current version of the SDK, users match the `Unhandled` variant. They can then read the code from the `Unhandled` variant because [`Unhandled`](https://docs.rs/aws-smithy-types/0.56.1/aws_smithy_types/error/struct.Unhandled.html) implements the `ProvideErrorMetadata` trait as well as the standard-library `std::error::Error` trait.
+
+> Note: It's possible to write correct code today because the operation-level and service-level errors already expose `code()` via `ProvideErrorMetadata`. This RFC describes mechanisms to guide customers to write forward-compatible code.
+
+```rust,ignore
+# fn docs() {
+    match client.get_object().send().await {
+        Ok(obj) => { ... },
+        Err(e) => match e.into_service_error() {
+            GetObjectError::NotFound => { ... },
+            GetObjectError::Unhandled(err) if err.code() == "ValidationException" => { ... }
+            other => { /** do something with this variant */ }
+        }
+    }
+# }
+```
+
+We must instead guide customers into the following pattern:
+```rust,ignore
+# fn docs() {
+    match client.get_object().send().await {
+        Ok(obj) => { ... },
+        Err(e) => match e.into_service_error() {
+            GetObjectError::NotFound => { ... },
+            err if err.code() == "ValidationException" => { ... },
+            err => warn!("{}", err.code()),
+        }
+    }
+# }
+```
+
+In this example, because customers are _not_ matching on the `Unhandled` variant explicitly this code is forward compatible for `ValidationException` being introduced in the future.
+
+**Guiding Customers to this Pattern**
+There are two areas we need to handle:
+1. Prevent customers from extracting useful information from `Unhandled`
+2. Alert customers _currently_ using unhandled what to use instead. For example, the following code is still problematic:
+    ```rust,ignore
+        match err {
+            GetObjectError::NotFound => { ... },
+            err @ GetObject::Unhandled(_) if err.code() == Some("ValidationException") => { ... }
+        }
+    ```
+
+For `1`, we need to remove the `ProvideErrorMetadata` trait implementation from `Unhandled`. We would expose this isntead through a layer of indirection to enable code generated to code to still read the data.
+
+For `2`, we would deprecate the `Unhandled` variants with a message clearly indicating how this code should be written.
```

Comment:
We should document here that this is different from how we do `Unknown` on enums because of the `ProvideErrorMetadata` trait. There's no equivalent for the generated enums.

Also, should we consider deprecating `Unknown` in the enum generator for the same reason? It would at least guide customers in the right direction better since right now it's just completely useless, and thus, confusing.

Reply:
> Also, should we consider deprecating Unknown in the enum generator for the same reason?
absolutely. I was thinking the same thing. We can also use it to centralize some of the docs around that